### PR TITLE
DeviceTypesClient implemented.

### PR DESCRIPTION
**Pull request by SezginEge**

### CODE REVIEW

File: xomni-sdk-ts-preview-test/fixtures/management/company/device-types-fixture.ts [hunk 1]
```diff
@@ -0,0 +1,492 @@
...
+            for (var i = 0; i < 200; i++) {
+                invalidDescription += "x";
+            }
+            var invalidDeviceType = <Models.Management.Company.DeviceType> {
+                Id: validId,
+                Description: invalidDescription
+            };
+
+            testClient.post(invalidDeviceType, suc => { }, err => { })
+        }).toThrow(new Error("description could not be greater than 150"));
```
Comment (ilkayilknur) [inline xomni-sdk-ts-preview-test/fixtures/management/company/device-types-fixture.ts]: "description could not be greater than 150" doesn't mean anything to caller. This message needs to be "description length could not be greater than 150". Could you please update this message ?

File: xomni-sdk-ts-preview/src/management/company/device-types.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+            Xomni.Utils.Validator.isDefined("deviceType", deviceType);
+            Xomni.Utils.Validator.isDefined("description", deviceType.Description);
+            Xomni.Utils.Validator.isLessThan(deviceType.Description.length, "description", 150);
+
+            this.httpProvider.post(this.baseUri, deviceType, success, error);
+        }
+
+        put(deviceType: Models.Management.Company.DeviceType, success: (result: Models.Management.Company.DeviceType) => void, error: (error: Models.ExceptionResult) => void) {
+            Xomni.Utils.Validator.isDefined("deviceType", deviceType);
+            Xomni.Utils.Validator.isGreaterThanOrEqual("id", deviceType.Id, 0);
```
Comment (ilkayilknur) [inline xomni-sdk-ts-preview/src/management/company/device-types.ts]: deviceTypeId could not be equal to 0. You need to change validation to Xomni.Utils.Validator.isGreaterThanOrEqual("id", deviceType.Id, 1);

File: xomni-sdk-ts-preview/src/management/company/device-types.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+            Xomni.Utils.Validator.isDefined("description", deviceType.Description);
+            Xomni.Utils.Validator.isLessThan(deviceType.Description.length, "description", 150);
+
+            this.httpProvider.put(this.baseUri, deviceType, success, error);
+        }
+
+        delete(deviceTypeId: number, success: () => void, error: (error: Models.ExceptionResult) => void) {
+            Xomni.Utils.Validator.isGreaterThanOrEqual("deviceTypeId", deviceTypeId, 0);
```
Comment (ilkayilknur) [inline xomni-sdk-ts-preview/src/management/company/device-types.ts]: Please see my previous comment.

File: xomni-sdk-ts-preview/src/management/company/device-types.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+            var uri = Xomni.Utils.UrlGenerator.PrepareOperationUrl(this.baseUri, deviceTypeId.toString());
+
+            this.httpProvider.delete(uri, success, error);
+        }
+
+        get(deviceTypeId: number, success: (result: Models.Management.Company.DeviceType) => void, error: (error: Models.ExceptionResult) => void) {
+            Xomni.Utils.Validator.isGreaterThanOrEqual("deviceTypeId", deviceTypeId, 0);
```
Comment (ilkayilknur) [inline xomni-sdk-ts-preview/src/management/company/device-types.ts]: Same here. DeviceTypeId could not be equal to 0.

File: xomni-sdk-ts-preview/src/management/company/device-types.ts [hunk 5]
```diff
@@ -0,0 +1,48 @@
+﻿/// <reference path="../../xomni.ts" />
+module Xomni.Management.Company.DeviceTypes {
+    export class DeviceTypesClient extends BaseClient {
+        private baseUri: string = "/management/company/devicetypes";
+
+        post(deviceType: Models.Management.Company.DeviceType, success: (result: Models.Management.Company.DeviceType) => void, error: (error: Models.ExceptionResult) => void) {
+            Xomni.Utils.Validator.isDefined("deviceType", deviceType);
+            Xomni.Utils.Validator.isDefined("description", deviceType.Description);
+            Xomni.Utils.Validator.isLessThan(deviceType.Description.length, "description", 150);
+
```
Comment (ilkayilknur) [inline xomni-sdk-ts-preview/src/management/company/device-types.ts]: These validation codes are also duplicating in put method. Maybe you should consider wrapping these validation codes to in a new method. Calling this method in post&put methods.

File: xomni-sdk-ts-preview/src/management/company/device-types.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+            Xomni.Utils.Validator.isGreaterThanOrEqual("deviceTypeId", deviceTypeId, 0);
+            var uri = Xomni.Utils.UrlGenerator.PrepareOperationUrl(this.baseUri, deviceTypeId.toString());
+
+            this.httpProvider.get(uri, success, error);
+        }
+
+        getList(skip: number, take: number, succes: (result: Models.PaginatedContainer<Models.Management.Company.DeviceType>) => void, error: (error: Models.ExceptionResult) => void) {
+            Xomni.Utils.Validator.isGreaterThanOrEqual("skip", skip, 0);
+            Xomni.Utils.Validator.isGreaterThanOrEqual("take", take, 1);
+            var uri = Xomni.Utils.UrlGenerator.PrepareOperationUrlWithMultipleParameter(this.baseUri, new Dictionary<string, string>([
```
Comment (ilkayilknur) [inline xomni-sdk-ts-preview/src/management/company/device-types.ts]: Method name needs to be PrepareOperationUrlWithMultipleParameter**s**.

File: xomni-sdk-ts-preview/src/management/company/device-types.ts (hunk 2 above)
Comment (SezginEge) [inline xomni-sdk-ts-preview/src/management/company/device-types.ts]: @gokhangulbiz told me that deviceTypeId must be greater than or equal to 0. Are you sure @ilkayilknur ?

Comment (ilkayilknur) [inline xomni-sdk-ts-preview/src/management/company/device-types.ts]: yes.

**Comment by SezginEge** [comment]
@ilkayilknur bump!

**Comment by ilkayilknur** [comment]
@SezginEge Last week, I implemented something similar to public SDK. Can you take a look ? https://github.com/XomniCloud/xomni-sdk-dotnet/pull/9/files#diff-a440debd95a31a56a4b8369b2f0f3b94R22

**Comment by SezginEge** [comment]
This pr was closed. Because It needed to clean up. So, I sent a new pr #29 .

**Pull request closed without merging**